frontend/src/scenes/session-recordings/player/sessionEventsDataLogic.ts

```diff
@@ -0,0 +1,289 @@
+            (s) => [s.sessionEventsData],
+            (sessionEventsData): RecordingEventType[] =>
+                (sessionEventsData || []).filter((e) => e.event === '$web_vitals'),
+        ],
+        AIEvents: [
+            (s) => [s.sessionEventsData],
+            (sessionEventsData): RecordingEventType[] =>
+                // see if event start with $ai_
+                (sessionEventsData || []).filter((e) => e.event.startsWith('$ai_')),
+        ],
+        exceptionEvents: [
+            (s) => [s.sessionEventsData],
+            (sessionEventsData): RecordingEventType[] =>
+                (sessionEventsData || []).filter((e) => e.event === '$exception'),
+        ],
+        preloadableEvents: [
+            (s) => [s.webVitalsEvents, s.AIEvents, s.exceptionEvents],
+            (webVitalsEvents, AIEvents, exceptionEvents): RecordingEventType[] => [
+                ...webVitalsEvents,
+                ...AIEvents,
+                ...exceptionEvents,
+            ],
+            { resultEqualityCheck: objectsEqual },
+        ],
+        eventViewportsItems: [
+            (s) => [s.sessionEventsData],
+            (
+                sessionEventsData
+            ): TimeTree<{
+                timestamp: Dayjs
+                payload: ViewportResolution
+            }> => {
+                const viewportEvents = new TimeTree<{
+                    timestamp: Dayjs
+                    payload: ViewportResolution
+                }>()
+                viewportEvents.add(
+                    (sessionEventsData || [])
+                        .filter((e) => e.properties.$viewport_width && e.properties.$viewport_height)
+                        .map((e) => ({
+                            timestamp: dayjs(e.timestamp),
+                            payload: {
+                                width: e.properties.$viewport_width,
+                                height: e.properties.$viewport_height,
+                                href: e.properties.$current_url,
+                            },
+                        }))
+                )
+                return viewportEvents
+            },
+            { resultEqualityCheck: objectsEqual },
+        ],
+        viewportForTimestamp: [
+            (s) => [s.eventViewportsItems],
+            (eventViewportsItems) => {
+                return (timestamp: number) => {
+                    const closestItem =
+                        eventViewportsItems.next(dayjs(timestamp)) || eventViewportsItems.previous(dayjs(timestamp))
+                    if (!closestItem) {
+                        return undefined
+                    }
+                    return closestItem.payload as ViewportResolution
+                }
+            },
+        ],
+    })),
+    subscriptions(({ actions }) => ({
+        preloadableEvents: (pe: null | RecordingEventType[]) => {
+            if (pe?.length) {
+                actions.loadFullEventData(pe)
+            }
+        },
+    })),
```

Comment:
the reason for all of this splitting is that this function is pretty slow
i think subscriptions are inherently slower than actions/listeners

but figuring out how / where to change it was breaking my brain

splitting all of this up is way overdue